syntax = "proto3";

package example;

message SimpleMessage { int32 field = 1; }

message KeyValueMessage {
  string key = 1;
  int32 value = 2;
}

message MapMessage {
  repeated KeyValueMessage repeated_keyvalue_field = 1;
  map<string, int32> map_field = 2;
}

message ComplexMessage {
  enum Enum { ITEM_0 = 0; ITEM_1 = 1; ITEM_2 = 2; }

  double double_field = 1;
  float float_field = 2;
  int32 int32_field = 3;
  int64 int64_field = 4;
  uint32 uint32_field = 5;
  uint64 uint64_field = 6;
  sint32 sint32_field = 7;
  sint64 sint64_field = 8;
  fixed32 fixed32_field = 9;
  fixed64 fixed64_field = 10;
  sfixed32 sfixed32_field = 11;
  sfixed64 sfixed64_field = 12;
  bool bool_field = 13;
  string string_field = 14;
  bytes bytes_field = 15;
  Enum enum_field = 16;
  repeated int32 repeated_field = 17;
  SimpleMessage nested_field = 18;
  repeated SimpleMessage repeated_nested_field = 19;
  repeated KeyValueMessage repeated_keyvalue_field = 20;
  map<string, int32> simple_map_field = 21;
  map<string, MapMessage> complex_map_field = 22;
  repeated MapMessage repeated_map_field = 23;
}
